File: algorithms/single-number/javascript/single-number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var singleNumber = function (nums) {
  let map = new Map();

  nums.forEach((i) => {
    map.has(i) ? map.set(i, map.get(i) + 1) : map.set(i, 1);
  });

  for (const [key, value] of map) {
    if (value === 1) {
      return key;
    }
  }
};

File: algorithms/single-number/single-number.rs
impl Solution {
    pub fn single_number(nums: Vec<i32>) -> i32 {
        use std::collections::HashMap;
        let mut map = HashMap::<i32, i32>::new();

        for i in nums.into_iter() {
            if map.contains_key(&i) {
                *map.get_mut(&i).unwrap() += 1;
            } else {
                map.insert(i, 1);
            }
        }

        *map.iter()
            .find_map(|(key, &val)| if val == 1 { Some(key) } else { None })
            .unwrap()
    }
}
